From 3ace30d3e11385507de5b875668e6b377414149e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Tue, 1 Apr 2014 18:37:28 +0200
Subject: [PATCH] Fix race with metacity in fullscreen

To avoid some races with metacity, the window should be placed as
early as possible, before it is (re)allocated & mapped (rhbz#809546).
---
 src/virt-viewer-window.c | 25 ++++++++++++++-----------
 1 file changed, 14 insertions(+), 11 deletions(-)

diff --git a/src/virt-viewer-window.c b/src/virt-viewer-window.c
index 860cfc1..725914e 100644
--- a/src/virt-viewer-window.c
+++ b/src/virt-viewer-window.c
@@ -441,7 +441,7 @@ virt_viewer_window_move_to_monitor(VirtViewerWindow *self)
     GdkRectangle mon;
     gint n = priv->fullscreen_monitor;
 
-    if (n == -1 || !priv->fullscreen)
+    if (n == -1)
         return;
 
     gdk_screen_get_monitor_geometry(gdk_screen_get_default(), n, &mon);
@@ -510,19 +510,25 @@ virt_viewer_window_enter_fullscreen(VirtViewerWindow *self, gint monitor)
     priv->fullscreen_monitor = monitor;
 
     if (!gtk_widget_get_mapped(priv->window)) {
+        /*
+         * To avoid some races with metacity, the window should be placed
+         * as early as possible, before it is (re)allocated & mapped
+         * Position & size should not be queried yet. (rhbz#809546). 
+         */
+        virt_viewer_window_move_to_monitor(self);
         g_signal_connect(priv->window, "map-event", G_CALLBACK(mapped), self);
         return;
+    } else {
+        gtk_window_get_position(GTK_WINDOW(priv->window),
+                                &priv->before_fullscreen.x,
+                                &priv->before_fullscreen.y);
+        gtk_window_get_size(GTK_WINDOW(priv->window),
+                            &priv->before_fullscreen.width,
+                            &priv->before_fullscreen.height);
     }
 
     priv->fullscreen = TRUE;
 
-    gtk_window_get_position(GTK_WINDOW(priv->window),
-                            &priv->before_fullscreen.x,
-                            &priv->before_fullscreen.y);
-    gtk_window_get_size(GTK_WINDOW(priv->window),
-                        &priv->before_fullscreen.width,
-                        &priv->before_fullscreen.height);
-
     gtk_check_menu_item_set_active(check, TRUE);
     gtk_widget_hide(menu);
     gtk_widget_show(priv->toolbar);
@@ -533,7 +539,6 @@ virt_viewer_window_enter_fullscreen(VirtViewerWindow *self, gint monitor)
         virt_viewer_display_set_monitor(priv->display, monitor);
         virt_viewer_display_set_fullscreen(priv->display, TRUE);
     }
-    virt_viewer_window_move_to_monitor(self);
 
     gtk_window_fullscreen(GTK_WINDOW(priv->window));
 
@@ -1167,8 +1172,6 @@ virt_viewer_window_show(VirtViewerWindow *self)
         virt_viewer_window_resize(self, FALSE);
         self->priv->desktop_resize_pending = FALSE;
     }
-
-    virt_viewer_window_move_to_monitor(self);
 }
 
 void
